eng/scripts/Set-TemplateTestMatrixVariable.ps1

```diff
@@ -0,0 +1,97 @@
+<#
+.SYNOPSIS
+Generates a matrix of template test jobs.
+
+.PARAMETER TemplateList
+List of templates to run. By default, uses `azd template list` to run all templates.
+
+.PARAMETER TemplateListFilter
+Regex filter expression to filter templates. Examples: 'csharp', 'terraform', 'python-mongo'.
+
+.PARAMETER JobVariablesDefinition
+Job variables that will be set on each matrix job, in the format of a comma-delimited list of key=value.
+
+.PARAMETER OutputMatrixVariable
+The output variable that will contain the matrix job definitions.
+The matrix job definition will contain:
+- TemplateName - the name of the template being tested
+- UseUpperCase - whether an upper-case version of the template name should be tested
+- AzdContainerImage - the container image for the template test
+- Any job variables defined in JobVariablesDefinition
+
+#>
+param (
+    # This is a string and not a string[] to avoid issues with parameter passing in CI yaml.
+    [string]$TemplateList = '(azd template list)',
+    [string]$TemplateListFilter = '.*',
+    [string]$OutputMatrixVariable = 'Matrix',
+    [string]$JobVariablesDefinition = ''
+)
+
+function Get-JobVariables() {
+    param(
+        [string]$JobVariablesDefinition
+    )
+    $result = @{}
+    $definitions = ($JobVariablesDefinition -split ',').Trim()
+    if (-not $definitions) {
+        return $result
+    }
+
+    foreach ($definition in $definitions) {
+        $keyValue = ($definition -split '=').Trim()
+        if ($keyValue.Length -ne 2) {
+            throw "Invalid job variable definition: $definition"
+        }
+
+        $result[$keyValue[0]] = $keyValue[1]
+    }
+
+    return $result
+}
+
+$jobVariables = Get-JobVariables -JobVariablesDefinition $JobVariablesDefinition
+
+if ($TemplateList -eq '(azd template list)') {
+    Write-Host "Using results of (azd template list --output json)"
+    
+    $templateNames = (azd template list --output json | ConvertFrom-Json).name
+    if ($LASTEXITCODE -ne 0) {
+        Write-Error "azd template list failed"
+        exit 1
+    }
+} else {
+    Write-Host "Using provided TemplateList value: $TemplateList"
+
+    $templateNames = ($TemplateList -split ",").Trim()
+}
+
+if ($TemplateListFilter -ne '.*') {
+} else {
+    Write-Host "Filtering with TemplateListFilter regex: $TemplateListFilter"
+
+    $templateNames = $templateNames -match $TemplateListFilter
+}
+
+$matrix = @{}
+foreach ($template in $templateNames) {
+    $jobName = $template.Replace('/', '_')
+    $matrix[$jobName] = @{ TemplateName = $template }
+}
+
+# Adding extra test for capitals letters support. Using first template
+$firstTemplate = $templateNames[0]
+$capitalsTest = $firstTemplate.Replace('/', '_') + "-Upper-case-test"
+$matrix[$capitalsTest] = @{ TemplateName = $firstTemplate; UseUpperCase = "true" }
```

Comment:
I know you are just copying/re-organizing this code, but as I'm looking at it for the first time, I'm curious if you or @danieljurek know the reason why we need to add tests via a powershell script as opposed to just defining extras in the matrix file and using the `-Filter` or `-Replace` flags to tune what gets tested?

Reply:
@benbp This was resolved as part of the refactoring PR, but happy to take suggestions to improve this as part of another PR in the future. The extra upper-case-test really should be redone differently -- probably as part of the CLI validation and not template test validation.

Reply:
I think CLI validation makes sense as you say.

Reply:
But my general approach with matrix test formulation is to try to push as much generation of tests into the matrix itself, and use the existing parameters to filter/update the matrix if needed with moderation. This way it's easier to look at a matrix config and reason about what test cases will actually run, vs. having to parse through a lot of yaml/scripts to figure out what will be tested (and it's harder to reproduce the same test runs/matrix cases locally).